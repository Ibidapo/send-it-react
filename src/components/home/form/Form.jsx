import React, { Component } from 'react';
import { isEmail } from 'validator';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';

import toastOptions from '../../../utils/toastOptions';
import './styles.scss';

class Form extends Component {
  state = {
    email: "",
    password: "",
    disableButton: false,
    displaySpinner: false,
  }

  onEmailChange = ({ target: { value } }) => this.setState(() => ({ email: value }));
  onPasswordChange = ({ target: { value } }) => this.setState(() => ({ password: value }));

  onSubmit = (e) => {
    e.preventDefault();
    const { props: { onSubmit }, state: { email, password } } = this;

    if (!isEmail(email)) return toast.error('Email address is not valid', toastOptions);
    if (password.length < 4) 
      return toast.error('Password must be greater than 4 characters', toastOptions);

    onSubmit({ email, password});
  }

  render() {
    const { 
      state: { email, password }, props: { button_label },
      onEmailChange, onPasswordChange, onSubmit
    } = this;

    return (
      <form id="form" className="mx-auto" onSubmit={onSubmit}>
        <div className='pb-1'>
          <label htmlFor="form-email">Email</label>
          <input
            type="text"
            value={email}
            onChange={onEmailChange}
          />
        </div>
        <div className="py-1">
          <label htmlFor="form-pass">Password</label>
            <input
              type="password"
              value={password}
              onChange={onPasswordChange}
            />
        </div>
        <div className="text-center pt-1">
          <button id="form-btn">{button_label}</button>
        </div>  
      </form>
    );
  }
}

Form.propTypes = {
  onSubmit: PropTypes.func,
  button_label: PropTypes.string
}

export default Form;