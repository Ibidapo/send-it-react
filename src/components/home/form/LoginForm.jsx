import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';

import Form from './Form';
import { loginUser } from '../../../redux/actions/auth';
import { getProfile } from '../../../redux/actions/profile';
import toastOptions from '../../../utils/toastOptions';

const LoginForm = ({ isLoggedIn, loginUser, getProfile, history }) => {
  const onSubmit = async ({ email, password }) => {
    if (isLoggedIn) return history.push('/dashboard');

    const { error, success, user } = await loginUser({ email, password });

    if (error) return toast.error(error, toastOptions);

    await getProfile(user)
    history.push('/dashboard');
    toast.success(success, toastOptions);
  }

  return (
    <Form button_label='Login' onSubmit={onSubmit} />
  );
}

LoginForm.propTypes = {
  isLoggedIn: PropTypes.bool,
  loginUser: PropTypes.func,
  history: PropTypes.object,
  getProfile: PropTypes.func
}

const mapStateToProps = ({ auth: { isLoggedIn }, profile }) => ({ isLoggedIn, profile });
const mapDispatchToProps = ({ loginUser, getProfile });

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);