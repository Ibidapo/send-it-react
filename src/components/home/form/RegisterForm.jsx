import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';

import Form from './Form';
import { registerUser } from '../../../redux/actions/auth';
import { getProfile } from '../../../redux/actions/profile';
import toastOptions from '../../../utils/toastOptions';

export const RegisterForm = ({ isLoggedIn, registerUser, getProfile, history }) => {
  const onSubmit = async ({ email, password }) => {
    if (isLoggedIn) return history.push('/dashboard');

    const { error, success, user } = await registerUser({ email, password });

    if (error) return toast.error(error, toastOptions);
    await getProfile(user)
    history.push('/dashboard');
    toast.success(success, toastOptions);
  }

  return (
    <Form button_label='Register' onSubmit={onSubmit} />
  );
}

RegisterForm.propTypes = {  
  isLoggedIn: PropTypes.bool,
  registerUser: PropTypes.func,
  history: PropTypes.object,
  getProfile: PropTypes.func
}

const mapStateToProps = ({ auth: { isLoggedIn }, profile }) => ({ isLoggedIn, profile });
const mapDispatchToProps = ({ registerUser, getProfile });

export default connect(mapStateToProps, mapDispatchToProps)(RegisterForm);