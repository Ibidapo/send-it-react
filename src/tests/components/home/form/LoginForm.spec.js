import React from 'react';
import { shallow } from 'enzyme';

import Form from '../../../../components/home/form/Form';
import { LoginForm } from '../../../../components/home/form/LoginForm';

let wrapper;

describe('Test for LoginForm Component', () => {
  it('Should render LoginForm Component', () => {
    const props = { isLoggedIn: false, loginUser: jest.fn(), getProfile: jest.fn(), history: { push: jest.fn() } }
    wrapper = shallow(<LoginForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  })

  it('Should submit form', () => {
    expect(wrapper.find(Form).length).toBe(1);
    wrapper.find(Form).at(0).simulate('submit', { preventDefault: () => { } });
  });

  it('Should submit form when logged in', () => {
    const props = { 
      isLoggedIn: true,
      loginUser: jest.fn(() => Promise.resolve({})), 
      getProfile: jest.fn(), 
      history: { push: jest.fn() } 
    }

    wrapper = shallow(<LoginForm {...props} />);
    wrapper.find(Form).at(0).simulate('submit', { preventDefault: () => { } });
  })
});