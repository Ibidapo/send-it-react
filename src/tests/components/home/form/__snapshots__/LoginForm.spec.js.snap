// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for LoginForm Component Should render LoginForm Component 1`] = `
<Form
  button_label="Login"
  onSubmit={[Function]}
/>
`;
